tools/render-test/render-d3d12.cpp

```diff
@@ -274,71 +1929,466 @@ ShaderCompiler* D3D12Renderer::getShaderCompiler()
+                break;
+            }
+            case ShaderInputType::CombinedTextureSampler:
+            {
+                assert(!"Not implemented");
+                //throw "not implemented";
+                return nullptr;
+                break;
+            }
+        }
+    }
+
+    return bindingState.detach();
 }
 
 void D3D12Renderer::setBindingState(BindingState* state)
 {
+    m_boundBindingState = static_cast<BindingStateImpl*>(state);
 }
 
-void D3D12Renderer::serializeOutput(BindingState* state, const char* fileName)
+void D3D12Renderer::serializeOutput(BindingState* stateIn, const char* fileName)
 {
+    auto bindingState = static_cast<BindingStateImpl*>(stateIn);
+    FILE * f = fopen(fileName, "wb");
+
+    D3D12_HEAP_PROPERTIES heapProps;
+    heapProps.Type = D3D12_HEAP_TYPE_READBACK;
+    heapProps.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
+    heapProps.MemoryPoolPreference = D3D12_MEMORY_POOL_UNKNOWN;
+    heapProps.CreationNodeMask = 1;
+    heapProps.VisibleNodeMask = 1;
+
+    int id = 0;
+    for (auto & binding : bindingState->m_bindings)
+    {
+        if (binding.m_isOutput)
+        {
+            if (binding.m_resource.getResource())
+            {
+                // create staging buffer
+
+                size_t bufferSize = D3DUtil::calcAligned(binding.m_bufferLength, 256);
+
+                D3D12_RESOURCE_DESC stagingDesc;
+                _initBufferResourceDesc(bufferSize, stagingDesc);
+                
+                D3D12Resource stageBuf;
+                SLANG_RETURN_VOID_ON_FAIL(stageBuf.initCommitted(m_device, heapProps, D3D12_HEAP_FLAG_NONE, stagingDesc, D3D12_RESOURCE_STATE_COPY_DEST, nullptr));
+                
+                const D3D12_RESOURCE_STATES initialState = binding.m_resource.getState();
+
+                // Make it a source
+                {
+                    D3D12BarrierSubmitter submitter(m_commandList);
+                    binding.m_resource.transition(D3D12_RESOURCE_STATE_COPY_SOURCE, submitter);
+                }
+                // Do the copy
+                m_commandList->CopyBufferRegion(stageBuf, 0, binding.m_resource, 0, bufferSize);
+                // Switch it back
+                {
+                    D3D12BarrierSubmitter submitter(m_commandList);
+                    binding.m_resource.transition(initialState, submitter);
+                }
+
+                // Wait until complete
+                submitGpuWorkAndWait();
+
+                UINT8* data;
+                D3D12_RANGE readRange = {0, bufferSize};
+                
+                SLANG_RETURN_VOID_ON_FAIL(stageBuf.getResource()->Map(0, &readRange, reinterpret_cast<void**>(&data)));
+                {
+                    auto ptr = (unsigned int *)data;
+                    for (auto i = 0u; i < binding.m_bufferLength / sizeof(unsigned int); i++)
+                        fprintf(f, "%X\n", ptr[i]);
```

Comment:
Aside: This logic in particular is one of the places that I'd like to hoist out of the different `Renderer` implementations and into a central place. We could probably have the tests do a much better job of formatting the output (e.g., to handle both float and integer types nicely), but that would currently involve duplication at all the call sites that currently do this same `fprintf`.